src/KawaBot.Server/HttpListener.fs

```diff
@@ -14,43 +14,55 @@ module Http =
     type HttpListener =
         val private _url: string
         val private _handlers: Dictionary<string, UrlHandler>
+        val private _handleNotFound: HandlerFunc
+        
         [<DefaultValue>]
-        val mutable private _isRunning: bool
+        val mutable private _listener: System.Net.HttpListener
 
         new(url: string) =
             {
-                _url = url;
-                _handlers = Dictionary();
+                _url = url
+                _handlers = Dictionary()
+                _handleNotFound = fun req resp -> async {
+                    resp.StatusCode <- (int)HttpStatusCode.NotFound
+                }
             }
 
         member this.Register(handler: UrlHandler): unit =
             this._handlers.Add(handler.Url, handler);
             
         member this.Run(): unit = 
-            this._isRunning <- true
-            let listener = new System.Net.HttpListener()
-            listener.Prefixes.Add this._url
-            listener.Start()
+            this._listener <- new System.Net.HttpListener()
+            this._listener.Prefixes.Add this._url
+            this._listener.Start()
             async {
                 printfn "[%s] Listening to %s started" (Utils.ToDateTimeString DateTime.Now) this._url
                 printfn "[%s] Press any key to exit..." (Utils.ToDateTimeString DateTime.Now)
-                while this._isRunning do 
-                    let! context = Async.FromBeginEnd(listener.BeginGetContext, listener.EndGetContext)
-                    let url = context.Request.RawUrl.Substring(1).Split('?').[0];
-                    match this._handlers.TryGetValue(url) with
-                        | true, handler -> this.Invoke(handler, context) |> Async.Start
-                        | _ -> ()
+                let mutable keepRunning = true
+                while keepRunning do
+                    let optionalContext =
+                        try
+                            Some(this._listener.GetContext())
+                        with
+                            | :? HttpListenerException -> keepRunning <- false; None
```

Comment:
Don't we mute a potentially meaning exception? Why can't we use cancellation flag checking + checking [HttpListener.IsListening Property](https://msdn.microsoft.com/en-us/library/system.net.httplistener.islistening(v=vs.110).aspx) approach?